README.md
# zepto.fullpage 
[![release](https://img.shields.io/badge/release-v0.5.0-orange.svg)](https://github.com/yanhaijing/zepto.fullpage/releases) [![issues](https://img.shields.io/github/issues/yanhaijing/zepto.fullpage.svg)](https://github.com/yanhaijing/zepto.fullpage/issues) [![forks](https://img.shields.io/github/forks/yanhaijing/zepto.fullpage.svg)](https://github.com/yanhaijing/zepto.fullpage/network) [![stars](https://img.shields.io/github/stars/yanhaijing/zepto.fullpage.svg)](https://github.com/yanhaijing/zepto.fullpage/stargazers) [![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/yanhaijing/zepto.fullpage/blob/master/LICENSE)

专注于移动端的fullPage.js。

## 功能概述
可实现移动端的单页滚动效果，可自定义参数，提供回调接口，和公开接口。

## 兼容性
- Ios4+
- Andriod2.3+（未全部覆盖）

## 快速上手
### HTML

	<div class="wp">
        <div class="wp-inner">
            <div class="page page1">1</div>
            <div class="page page2">2</div>
            <div class="page page3">3</div>
            <div class="page page4">4</div>
        </div>
    </div>

### CSS
父容器需是固定高度，并且溢出为隐藏，fullpage会使用父元素的宽度和高度。

	.wp{
	    overflow: hidden;            
	}

## js
一行代码即可完成，如此简单：

	$('.wp-inner').fullpage();

**注意：是在.wp-inner上调用的。**

更多例子，请移步[这里](demo)。

## 文档
[API](doc/api.md)

## 贡献指南
如果你想为zepto.fullpage贡献代码，请采用fork + pull request 方式，并在发起pr前先将master上超前的代码rebase到自己的分支上。

### 发布Bower
    
    $ bower register zepto.fullpage git://github.com/yanhaijing/zepto.fullpage.git

## 贡献者
- [yanhaijing](http://yanhaijing.com)
- [YuyingWu](http://blog.wuyuying.com/)

## 贡献者Q群
如果你有任何疑问，请到[Issues](https://github.com/yanhaijing/zepto.fullpage/issues)反馈，此Q群为贡献者使用，仅供贡献者内部讨论使用。

点击加入：[zepto.fullpage官方交流群](http://jq.qq.com/?_wv=1027&k=bc6Grs)(314819079)

## 为什么会有这个项目
最近单页滚动的效果非常流行，这种效果的视觉冲击感很强烈，特别是对于活动页面，因为需求的迫切，在pc端诞生了 fullPage.js和onepage-scroll这样的库，在移动的浪潮下，我们自然需要把这种效果搬到移动设备上，但当我在想实现类似效果时，却找不到一个这样的库（fullpage.js对于移动端来说太大了，还要依赖jquery），所以我就开发了一个，先是用于自己项目中，后来抽了出来，就有了这个项目。

## 更新日志
[更新日志](CHANGELOG.md)

## TODO
[计划列表](TODO.md)

如果你有好的建议，欢迎[反馈给我](https://github.com/yanhaijing/zepto.fullpage/issues/42)。

## 谁在使用
- [百度知道](http://zhidao.baidu.com/)
- [百度经验](http://jingyan.baidu.com/)
- [百度师傅](http://shifu.baidu.com/)
- [拇指医生](http://muzhi.baidu.com/)
- [墨迹天气](http://www.mojichina.com/)
- [58同城](http://58.com)

**注:如果您也在使用，欢迎[反馈给我](https://github.com/yanhaijing/zepto.fullpage/issues/9)**

## 成功案例
- [百度知道-APP5.0落地页](http://zhidao.baidu.com/s/5_0-page/index.html)
- [重庆华西妇产医院-妈妈我来啦](http://d.cqhxfk.com/game/0505/)
- [百度经验-关于表白这事儿](http://jingyan.baidu.com/z/2015-520/index.html)

更多案例，请点击[这里查看](https://github.com/yanhaijing/zepto.fullpage/issues/9)。

## 相关链接
- [fullPage.js](https://github.com/alvarotrigo/fullPage.js)
- [onepage-scroll](https://github.com/peachananr/onepage-scroll)
- [zepto-onepage-scroll](https://github.com/peachananr/zepto-onepage-scroll)
- [parallax.js](https://github.com/hahnzhu/parallax.js)
- [H5FullscreenPage](https://github.com/lvming6816077/H5FullscreenPage)
- [fullpage](https://github.com/powy1993/fullpage)

如果你知道其他第三方库，欢迎[反馈给我](https://github.com/yanhaijing/zepto.fullpage/issues/50)。

## 自动化工具
- [爆米兔移动营销平台](http://www.baomitu.com/)
- [数说 · Legend(度厂内部)](http://legend.baidu.com/) 

如果你知道其他第三方工具，欢迎[反馈给我](https://github.com/yanhaijing/zepto.fullpage/issues/51)。



CHANGELOG.md
# 0.5.0 / 2016/1/26
- 添加holdTouch/unholdTouch 方法
- 修改start/stop 方法的逻辑
- 添加使用 animations.css 的demo
- 添加播放声音的demo
- 添加一种小屏幕解决方案

# 0.4.0 / 2016/1/23
- 使用translate3d硬件加速
- 首尾屏添加拖动阻力
- 添加纯js实现版本，无zepto依赖
- 增加滑动阈值判断
- 增加获取当前是第几屏的方法
- 扩展beforeChange返回值的功能，若beforeChange显示返回false，则可阻止滚屏的发生
- 添加丰富的demo
- 修复文档描述错误

# 0.3.1 / 2015/5/20
- 修改change and afterChange 默认参数
- 修复阻止touch事件的默认行为的bug
- add file meta
- 提升性能

# 0.3.0 / 2015.05.05
- 添加水平切换功能
- 提升性能

# 0.2.0 / 2015.04.01
- 新增循环滚动功能
- 新增start stop方法
- 优化性能，使用transform代替top left

# 0.1.0 / 2015.02.11
- 单页滚动功能
- 自定义功能
- 丰富的回调函数
- 丰富的公开方法
- 尺寸变更，自动重渲染
- CSS 3D加速功能


src/js.fullpage.js
/*!
 * zepto.fullpage.js v0.4.0 (https://github.com/yanhaijing/zepto.fullpage)
 * API https://github.com/yanhaijing/zepto.fullpage/blob/master/doc/api.md
 * Copyright 2014 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/yanhaijing/zepto.fullpage/blob/master/LICENSE)
 */
(function(window) {
    var d = {
        page: '.page',
        start: 0,
        duration: 500,
        loop: false,
        drag: false,
        dir: 'v',
        der: 0.1,
        change: function(data) {},
        beforeChange: function(data) {},
        afterChange: function(data) {},
        orientationchange: function(orientation) {}
    };

    function touchmove(e) {
        e.preventDefault();
    }

    function fix(cur, pagesLength, loop) {
        if (cur < 0) {
            return !!loop ? pagesLength - 1 : 0;
        }

        if (cur >= pagesLength) {
            return !!loop ? 0 : pagesLength - 1;
        }

        return cur;
    }

    function move(ele, dir, dist) {
        var xPx = "0px" , yPx = "0px";
        if(dir === 'v') yPx = dist+"px";
        else xPx = dist + "px";
        ele.style.cssText += (';-webkit-transform : translate3d(' + xPx + ', ' + yPx + ', 0px);' +
                                'transform : translate3d(' + xPx + ', ' + yPx + ', 0px);');
    }

    function init(option) {
        var o = option ? option : {};
        for (var key in d) {
            if(!o.hasOwnProperty(key)){
                o[key] = d[key];
            }
        }

        var that = this;
        that.curIndex = -1;
        that.o = o;

        that.startY = 0;
        that.movingFlag = false;

        that.ele.classList.add('fullPage-wp');
        that.parentEle = that.ele.parentNode;

        var query = o.page;
        if(query.indexOf(".")==0){
            query = query.substring(1, query.length);
        }
        that.pageEles = that.ele.getElementsByClassName(query);
        for (var i = 0; i < that.pageEles.length; i++) {
            var pageEle = that.pageEles[i];
            pageEle.classList.add('fullPage-page');
            pageEle.classList.add('fullPage-dir-' + o.dir);
        }

        that.pagesLength = that.pageEles.length;
        that.update();
        that.initEvent();
        that.start();
    }

    function Fullpage(ele, option) {
        this.ele = ele;
        init.call(this, option);
    }
    Fullpage.prototype.update = function() {
        if (this.o.dir === 'h') {
            this.width = this.parentEle.offsetWidth;
            for (var i = 0; i < this.pageEles.length; i++) {
                var pageEle = this.pageEles[i];
                pageEle.style.width = this.width+ "px";
            }
            this.ele.style.width = (this.width * this.pagesLength)+ "px";
        }

        this.height = this.parentEle.offsetHeight;
        for (var i = 0; i < this.pageEles.length; i++) {
            var pageEle = this.pageEles[i];
            pageEle.style.height = this.height + "px";
        }

        this.moveTo(this.curIndex < 0 ? this.o.start : this.curIndex);
    };

    Fullpage.prototype.initEvent = function() {
        var that = this;
        var ele = that.ele;

        ele.addEventListener('touchstart', function(e) {
            if (!that.status) {return 1;}
            //e.preventDefault();
            if (that.movingFlag) {
                return 0;
            }

            that.startX = e.targetTouches[0].pageX;
            that.startY = e.targetTouches[0].pageY;
        });
        ele.addEventListener('touchend', function(e) {
            if (!that.status) {return 1;}
            //e.preventDefault();
            if (that.movingFlag) {
                return 0;
            }

            var sub = that.o.dir === 'v' ? (e.changedTouches[0].pageY - that.startY)/that.height : (e.changedTouches[0].pageX - that.startX)/that.width;
            var der = (sub > that.o.der || sub < -that.o.der) ? sub > 0 ? -1 : 1 : 0;

            that.moveTo(that.curIndex + der, true);
        });
        if (that.o.drag) {
            ele.addEventListener('touchmove', function(e) {
                if (!that.status) {return 1;}
                //e.preventDefault();
                if (that.movingFlag) {
                    that.startX = e.targetTouches[0].pageX;
                    that.startY = e.targetTouches[0].pageY;
                    return 0;
                }

                var y = e.changedTouches[0].pageY - that.startY;
                if( (that.curIndex==0 && y>0) || (that.curIndex===that.pagesLength-1 && y<0) ) y/=2;
                var x = e.changedTouches[0].pageX - that.startX;
                if( (that.curIndex==0 && x>0) || (that.curIndex===that.pagesLength-1 && x<0) ) x/=2;
                var dist = (that.o.dir === 'v' ? (-that.curIndex * that.height + y) : (-that.curIndex * that.width + x));
                ele.classList.remove('anim');
                move(ele, that.o.dir, dist);
            });
        }

        // 翻转屏幕提示
        // ==============================
        window.addEventListener("orientationchange", function() {
            if (window.orientation === 180 || window.orientation === 0) {
                that.o.orientationchange('portrait');
            }
            if (window.orientation === 90 || window.orientation === -90) {
                that.o.orientationchange('landscape');
            }
        }, false);

        window.addEventListener("resize", function() {
            that.update();
        }, false);
    };
    Fullpage.prototype.holdTouch = function() {
        document.addEventListener('touchmove', touchmove);
    };
    Fullpage.prototype.unholdTouch = function() {
        document.removeEventListener('touchmove', touchmove);
    };
    Fullpage.prototype.start = function() {
        this.status = 1;
        this.holdTouch();
    };
    Fullpage.prototype.stop = function() {
        this.status = 0;
        this.unholdTouch();
    };
    Fullpage.prototype.getCurIndex = function () {
        return this.curIndex;
    };
    Fullpage.prototype.moveTo = function(next, anim) {
        var that = this;
        var ele = that.ele;
        var cur = that.curIndex;
        next = fix(next, that.pagesLength, that.o.loop);

        if (anim) {
            ele.classList.add('anim');
        } else {
            ele.classList.remove('anim');
        }

        if (next !== cur) {
            var flag = that.o.beforeChange({
                next: next,
                cur: cur
            });

            // beforeChange 显示返回false 可阻止滚屏的发生
            if (flag === false) {
                return 1;
            }
        }

        that.movingFlag = true;
        that.curIndex = next;
        move(ele, that.o.dir, -next * (that.o.dir === 'v' ? that.height : that.width));

        if (next !== cur) {
            that.o.change({
                prev: cur,
                cur: next
            });
        }

        window.setTimeout(function() {
            that.movingFlag = false;
            if (next !== cur) {
                that.o.afterChange({
                    prev: cur,
                    cur: next
                });
                for (var i = 0; i < that.pageEles.length; i++) {
                    var pageEle = that.pageEles[i];
                    if(i === next){
                        pageEle.classList.add("cur");
                    }else{
                        pageEle.classList.remove("cur");
                    }
                }
            }
        }, that.o.duration);
    };
    Fullpage.prototype.movePrev = function(anim) {
        this.moveTo(this.curIndex - 1, anim);
    };
    Fullpage.prototype.moveNext = function(anim) {
        this.moveTo(this.curIndex + 1, anim);
    };

    Element.prototype.fullpage = function(option) {
        return new Fullpage(this, option);
    };
}(window));


src/zepto.fullpage.js
/*!
 * zepto.fullpage.js v0.5.0 (https://github.com/yanhaijing/zepto.fullpage)
 * API https://github.com/yanhaijing/zepto.fullpage/blob/master/doc/api.md
 * Copyright 2014 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/yanhaijing/zepto.fullpage/blob/master/LICENSE)
 */
(function($, window, undefined) {
    if (typeof $ === 'undefined') {
        throw new Error('zepto.fullpage\'s script requires Zepto');
    }
    var fullpage = null;
    var d = {
        page: '.page',
        start: 0,
        duration: 500,
        loop: false,
        drag: false,
        dir: 'v',
        der: 0.1,
        change: function(data) {},
        beforeChange: function(data) {},
        afterChange: function(data) {},
        orientationchange: function(orientation) {}
    };

    function touchmove(e) {
        e.preventDefault();
    }
    
    function fix(cur, pagesLength, loop) {
        if (cur < 0) {
            return !!loop ? pagesLength - 1 : 0;
        }

        if (cur >= pagesLength) {
            return !!loop ? 0 : pagesLength - 1;
        }


        return cur;
    }

    function move($ele, dir, dist) {
        var xPx = '0px', yPx = '0px';
        if(dir === 'v') yPx = dist + 'px';
        else xPx = dist + 'px';
        $ele.css({
            '-webkit-transform' : 'translate3d(' + xPx + ', ' + yPx + ', 0px);',
            'transform' : 'translate3d(' + xPx + ', ' + yPx + ', 0px);'
        });
    }

    function init(option) {
        var o = $.extend(true, {}, d, option);
        var that = this;
        that.curIndex = -1;
        that.o = o;

        that.startY = 0;
        that.movingFlag = false;

        that.$this.addClass('fullPage-wp');
        that.$parent = that.$this.parent();
        that.$pages = that.$this.find(o.page).addClass('fullPage-page fullPage-dir-' + o.dir);
        that.pagesLength = that.$pages.length;
        that.update();
        that.initEvent();
        that.start();
    }

    function Fullpage($this, option) {
        this.$this = $this;
        init.call(this, option);
    }

    $.extend(Fullpage.prototype, {
        update: function() {
            if (this.o.dir === 'h') {
                this.width = this.$parent.width();
                this.$pages.width(this.width);
                this.$this.width(this.width * this.pagesLength);
            }

            this.height = this.$parent.height();
            this.$pages.height(this.height);

            this.moveTo(this.curIndex < 0 ? this.o.start : this.curIndex);
        },
        initEvent: function() {
            var that = this;
            var $this = that.$this;

            $this.on('touchstart', function(e) {
                if (!that.status) {return 1;}
                //e.preventDefault();
                if (that.movingFlag) {
                    return 0;
                }

                that.startX = e.targetTouches[0].pageX;
                that.startY = e.targetTouches[0].pageY;
            });
            $this.on('touchend', function(e) {
                if (!that.status) {return 1;}
                //e.preventDefault();
                if (that.movingFlag) {
                    return 0;
                }

                var sub = that.o.dir === 'v' ? (e.changedTouches[0].pageY - that.startY) / that.height : (e.changedTouches[0].pageX - that.startX) / that.width;
                var der = (sub > that.o.der || sub < -that.o.der) ? sub > 0 ? -1 : 1 : 0;

                that.moveTo(that.curIndex + der, true);
            });
            if (that.o.drag) {
                $this.on('touchmove', function(e) {
                    if (!that.status) {return 1;}
                    //e.preventDefault();
                    if (that.movingFlag) {
                        that.startX = e.targetTouches[0].pageX;
                        that.startY = e.targetTouches[0].pageY;
                        return 0;
                    }

                    var y = e.changedTouches[0].pageY - that.startY;
                    if( (that.curIndex == 0 && y > 0) || (that.curIndex === that.pagesLength - 1 && y < 0) ) y /= 2;
                    var x = e.changedTouches[0].pageX - that.startX;
                    if( (that.curIndex == 0 && x > 0) || (that.curIndex === that.pagesLength - 1 && x < 0) ) x /= 2;
                    var dist = (that.o.dir === 'v' ? (-that.curIndex * that.height + y) : (-that.curIndex * that.width + x));
                    $this.removeClass('anim');
                    move($this, that.o.dir, dist);
                });
            }

            // 翻转屏幕提示
            // ==============================             
            window.addEventListener('orientationchange', function() {
                if (window.orientation === 180 || window.orientation === 0) {
                    that.o.orientationchange('portrait');
                }
                if (window.orientation === 90 || window.orientation === -90) {
                    that.o.orientationchange('landscape');
                }
            }, false);

            window.addEventListener('resize', function() {
                that.update();
            }, false);
        },

        holdTouch: function() {
            $(document).on('touchmove', touchmove);
        },
        unholdTouch: function() {
            $(document).off('touchmove', touchmove);
        },
        start: function() {
            this.status = 1;
            this.holdTouch();
        },
        stop: function() {
            this.status = 0;
            this.unholdTouch();
        },
        moveTo: function(next, anim) {
            var that = this;
            var $this = that.$this;
            var cur = that.curIndex;

            next = fix(next, that.pagesLength, that.o.loop);

            if (anim) {
                $this.addClass('anim');
            } else {
                $this.removeClass('anim');
            }

            if (next !== cur) {
                var flag = that.o.beforeChange({
                    next: next,
                    cur: cur
                });

                // beforeChange 显示返回false 可阻止滚屏的发生
                if (flag === false) {
                    return 1;
                }
            }

            that.movingFlag = true;
            that.curIndex = next;
            move($this, that.o.dir, -next * (that.o.dir === 'v' ? that.height : that.width));

            if (next !== cur) {
                that.o.change({
                    prev: cur,
                    cur: next
                });
            }

            window.setTimeout(function() {
                that.movingFlag = false;
                if (next !== cur) {
                    that.o.afterChange({
                        prev: cur,
                        cur: next
                    });
                    that.$pages.removeClass('cur').eq(next).addClass('cur');
                }
            }, that.o.duration);

            return 0;
        },
        movePrev: function(anim) {
            this.moveTo(this.curIndex - 1, anim);
        },
        moveNext: function(anim) {
            this.moveTo(this.curIndex + 1, anim);
        },
        getCurIndex: function () {
            return this.curIndex;
        }
    });

    $.fn.fullpage = function(option) {
        if (!fullpage) {
            fullpage = new Fullpage($(this), option);
        }
        return this;
    };
    $.fn.fullpage.version = '0.5.0';
    //暴露方法
    $.each(['update', 'moveTo', 'moveNext', 'movePrev', 'start', 'stop', 'getCurIndex', 'holdTouch', 'unholdTouch'], function(key, val) {
        $.fn.fullpage[val] = function() {
            if (!fullpage) {
                return 0;
            }
            return fullpage[val].apply(fullpage, [].slice.call(arguments, 0));
        };
    });
}(Zepto, window));
